Cabal/src/Distribution/Simple/GHC/Build/Link.hs

```diff
@@ -76,128 +74,148 @@ linkOrLoadComponent
-            -- the one invocation, so that one has to include all the
-            -- linker stuff too, like -l flags and any .o files from C
-            -- files etc.
-            --
-            -- TODO: The repl doesn't use the runtime paths from linkerOpts
-            -- (ghcOptRPaths), which looks like a bug. After the refactor we
-            -- can fix this.
-            `mappend` linkerOpts mempty
-            `mappend` mempty
-              { ghcOptMode = toFlag GhcModeInteractive
-              , ghcOptOptimisation = toFlag GhcNoOptimisation
+                then extraLibsStatic bi
+                else extraLibs bi
+          , ghcOptLinkLibPath =
+              toNubListR $
+                if withFullyStaticExe lbi
+                  then cleanedExtraLibDirsStatic
+                  else cleanedExtraLibDirs
+          , ghcOptLinkFrameworks = toNubListR $ map getSymbolicPath $ PD.frameworks bi
+          , ghcOptLinkFrameworkDirs = toNubListR $ PD.extraFrameworkDirs bi
+          , ghcOptInputFiles =
+              toNubListR
+                [ coerceSymbolicPath $ buildTargetDir </> obj
+                | obj <- extraSourcesObjs
+                ]
+          , ghcOptNoLink = Flag False
+          , ghcOptRPaths = rpaths
+          }
+
+    case what of
+      BuildRepl replFlags -> liftIO $ do
+        let
+          -- For repl we use the vanilla (static) ghc options
+          staticOpts = buildOpts StaticWay
+          replOpts =
+            staticOpts
+              { -- Repl options use Static as the base, but doesn't need to pass -static.
+                -- However, it maybe should, for uniformity.
+                ghcOptDynLinkMode = NoFlag
+              , ghcOptExtra =
+                  Internal.filterGhciFlags
+                    (ghcOptExtra staticOpts)
+                    <> replOptionsFlags (replReplOptions replFlags)
+              }
+              -- For a normal compile we do separate invocations of ghc for
+              -- compiling as for linking. But for repl we have to do just
+              -- the one invocation, so that one has to include all the
+              -- linker stuff too, like -l flags and any .o files from C
+              -- files etc.
+              --
+              -- TODO: The repl doesn't use the runtime paths from linkerOpts
+              -- (ghcOptRPaths), which looks like a bug. After the refactor we
+              -- can fix this.
+              `mappend` linkerOpts mempty
+              `mappend` mempty
+                { ghcOptMode = toFlag GhcModeInteractive
+                , ghcOptOptimisation = toFlag GhcNoOptimisation
+                }
+          replOpts_final =
+            replOpts
+              { ghcOptInputModules = replNoLoad (replReplOptions replFlags) (ghcOptInputModules replOpts)
+              , ghcOptInputFiles = replNoLoad (replReplOptions replFlags) (ghcOptInputFiles replOpts)
               }
```

Comment:
At one stage there was an issue with `linkerOpts` also modifying the same fields so it seemed cleaner to separate out setting these options (which have to happen) rather than relying on whatever the semantics of `mappend` are.